HIP-3.md

```diff
@@ -0,0 +1,223 @@
+
+        feeQuotingIgp.payForGas(
+            _messageId,
+            _destinationDomain,
+            _ismAndHandleGas,
+            _refundAddress
+        );
+    }
+
+    function setIsmDestinationGas(uint32 _destinationDomain, uint256 _ismGas) external onlyOwner;
+}
+```
+
+The off-chain relayer relating to a `FeeQuotingIgp` will index `GasPayment` events and relate them to dispatched messages based off their message ID. Because the `FeeQuotingIgp` will revert if there was insufficient payment, all `GasPayment` events relate to messages that have fully paid for an amount of gas, and the relayer will honor these payments.
+
+Note that calldata costs for processing the message are not explicitly calculated by the fee quoting IGP. The maximum Hyperlane message size is 2 KiB, or 2048 bytes. Currently, the Ethereum calldata gas cost is 16 gas per non-zero byte, which comes out to a maximum of `32,768` gas. To avoid an additional required mapping in the InterchainGasCalculator of remote domain to the cost per byte of calldata, a flat cost of 33,000 gas to cover potential calldata costs is charged.
+
+### **Rationale**
+
+#### ISM-specific fee quoting IGPs
+
+On the origin chain, an IGP cannot know what the length of an arbitrary ISM's metadata calldata is, nor can it know the expected gas used by an arbitrary ISM's `verify` function. Therefore, the IGP must get this information from some source.
+
+Hyperlane users sending messages that use the default ISM should not need to worry about the destination gas costs incurred by the default ISM. This includes the metadata calldata and the `verify` function. Because the default ISM is out of user control and the parameters of the default ISM may change, a straightforward “happy path” for default ISM users is provided via the default ISM's IGP.
+
+Any non-default ISMs may also have their own ISM-specific IGPs created. Alternatively, users can always call the fee quoting IGP directly with their own `_ismAndHandleGas` amount.
+
+One initial idea was to provide the ISM-specific destination gas costs on the origin chain by having a view function on the local ISM:
+
+```
+function destinationGas(uint32 _destinationGas) external returns (uint256);
+```
+
+However, this conflicts with permissionless deployment of Hyperlane onto long-tail chains. A relayer can decide to offer its relaying services to process messages to long tail chains if they so choose, and it's more natural to put the onus on the relayer to specify what those costs are. 
+
+#### Calldata gas
+
+The gas used by the message's bytes in the calldata of a process transaction are not explicitly calculated by the InterchainGasPaymaster because the upper limit of the gas cost is low, and performing this calculation on-chain introduces additional storage reads and computational complexity. Additionally, providing the InterchainGasPaymaster with the exact size in bytes of the Hyperlane message isn't straightforward.
+
+#### Flexibility
```

Comment:
I'd still like to see a token amount only version of the interface, which IMO feels the most flexible

Reply:
Wondering how you feel about just saying that people can implement and use their own IGP with the same function interface but they could just not care about gas amount / refund address / any on-chain fee quoting like we did before https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/6046ce68ef32b986d4ed813ad656d05497911815/solidity/contracts/InterchainGasPaymaster.sol#L50